wasm/src/bits-builtins.c

```diff
@@ -0,0 +1,214 @@
+#include "arithmetic.h"
+#include "bits.h"
+#include "mpd.h"
+
+#define swap(x, y) { mpd_t *v = x; x = y; y = v; }
+
+opa_value *opa_bits_or(opa_value *a, opa_value *b)
+{
+    mpd_t *x = opa_number_to_bf(a);
+    mpd_t *y = opa_number_to_bf(b);
+    if (x == NULL || y == NULL)
+    {
+        opa_mpd_del(x);
+        opa_mpd_del(y);
+        return NULL;
+    }
+
+    if (mpd_sign(x) == mpd_sign(y))
+    {
+        if (mpd_sign(x)) // x neg
+        {
+            // (-x) | (-y) == ^(x-1) | ^(y-1) == ^((x-1) & (y-1)) == -(((x-1) & (y-1)) + 1)
+            mpd_t *x1 = qsub_one(qabs(x));
+            mpd_t *y1 = qsub_one(qabs(y));
+            mpd_t *z = qadd_one(qand(x1, y1));
+            return opa_bf_to_number(qneg(z));
+        }
+
+        // x | y == x | y
+        return opa_bf_to_number(qor(x, y));
+    }
+
+    // x.neg != y.neg
+	if (mpd_sign(x))
+    {
+        // | is symmetric
+        swap(x, y);
+	}
+
+	// x | (-y) == x | ^(y-1) == ^((y-1) &^ x) == -(^((y-1) &^ x) + 1)
+	mpd_t *y1 = qsub_one(qabs(y));
+    mpd_t *z = qadd_one(qand_not(y1, x));
+	return opa_bf_to_number(qneg(z));
+}
+
+opa_value *opa_bits_and(opa_value *a, opa_value *b)
+{
+    mpd_t *x = opa_number_to_bf(a);
+    mpd_t *y = opa_number_to_bf(b);
```

Comment:
I might have missed something, but what is going to happen if a float is passed in?

IIRC the golang implementation for these builtins will raise an error if a non-integer is supplied. If I read the implementation of `opa_number_to_bf` correctly it can handle getting a float. Should we add in a check for int only to match the golang implemenation?

Reply:
qor, qand, qxor and others should return NULL in that case. Now that you say this, I have to double check all the functions handle the NULL values properly without crashing when returning. At the moment, there's no other way to bail out from the builtins than the NULL return value...

Reply:
All the bits builtins should now return NULL if an integer parameter is not an integer.